pkg/ephemeral/io/tuple_streamer.go

```diff
@@ -0,0 +1,242 @@
+//
+// Copyright (c) 2022 - for information on the respective copyright owner
+// see the NOTICE file and/or the repository https://github.com/carbynestack/ephemeral.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+package io
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	"golang.org/x/sys/unix"
+	"math/big"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/carbynestack/ephemeral/pkg/castor"
+	. "github.com/carbynestack/ephemeral/pkg/types"
+)
+
+type PipeWriterFactory func(string, time.Duration) (PipeWriter, error)
+
+func DefaultPipeWriterFactory(filePath string, writeDeadline time.Duration) (PipeWriter, error) {
+	pr, err := NewTuplePipeWriter(filePath, writeDeadline)
+	return pr, err
+}
+
+type PipeWriter interface {
+	Write([]byte) (int, error)
+	Close() error
+}
+
+func NewTuplePipeWriter(filePath string, writeDeadline time.Duration) (*TuplePipeWriter, error) {
+	_ = os.Remove(filePath)
+	err := unix.Mkfifo(filePath, 0666)
+	if err != nil {
+		return nil, fmt.Errorf("Error creating pipe: %v\n", err)
+	}
+	file, err := os.OpenFile(filePath, os.O_WRONLY, os.ModeNamedPipe)
+	if err != nil {
+		return nil, fmt.Errorf("Error Opening File: %v\n", err)
+	}
+	return &TuplePipeWriter{
+		tupleFile:     file,
+		writeDeadline: writeDeadline,
+	}, nil
+}
+
+type TuplePipeWriter struct {
+	tupleFile     *os.File
+	writeDeadline time.Duration
+}
+
+func (tpw *TuplePipeWriter) Write(data []byte) (int, error) {
+	deadline := time.Now().Add(tpw.writeDeadline)
+	err := tpw.tupleFile.SetWriteDeadline(deadline)
+	if err != nil {
+		return 0, fmt.Errorf("Error setting write deadline: %v", err)
+	}
+	c, err := tpw.tupleFile.Write(data)
+	if err != nil {
+		return 0, fmt.Errorf("Error wrting data to pipe: %v", err)
+	}
+	return c, nil
+}
+
+func (tpw *TuplePipeWriter) Close() error {
+	return tpw.tupleFile.Close()
+}
+
+// TupleStreamer is an interface.
+type TupleStreamer interface {
+	StartStreamTuples(chan struct{}, chan error, *sync.WaitGroup)
+}
+
+const tupleBaseFolder = "Player-Data"
```

Comment:
made preprocessing data folder configurable using `ephemeral.spdz.prepFolder`

Reply:
Could you in that case also add some checks to parsing that config?
Other config parameters are validated in `cmd/ephemeral/main.go#InitTypedConfig`

Checks like
- Is the config value set at all?
- Does it point to an existing file/folder
- Does it point to a folder
- Maybe additionally: Is the structure inside that folder as expected (requires also checking the other config arguments)?

Reply:
* If config value is not set it will (indirectly) point to the working directory as it is used as relative path (as long as there is no leading `/`)
* folder structure will be created if it does not exist . See `spdz.go#363`
 ```go
 err := os.MkdirAll(playerDataDir, 0755)
 ```
* in case it is a file initialization will fail and panic on start
* the folder is expected to be empty. If any file already exists it will be removed / replaced anyways

Reply:
If the config value is not set, wouldn't the dir be put at root level then?
```golang
case castor.SpdzGfp:
		playerDataDir = fmt.Sprintf("%s/%d-%s-%d/",
			conf.PrepFolder, conf.PlayerCount, castor.SpdzGfp.Shorthand, conf.Prime.BitLen())
```

Not sure how nil is handled by the format here?
In any case, wouldn't it be better to have an explicit check (even if that's just setting the value to a default one) when creating the config?

Reply:
your're right, with `%s/%d-%s-%d/` and first parameter being an empty string will result in an absolute path at root level. However, `PrepFolder` should never be nil, as this is not a valid value (cannot bee used) for a string in go.

`Player-Data` is used as a default value when ephemeral is deployed using helm. 
I am not sure whether an non set `PrepFolder` should actually be considered an error case as system functionality is not affected as long as write access is granted (on root level) - not having write permissions on the defined folder is a failure case in general.